File: src/main/resources/static/templates/visualizations/scatterplot.js
//OpenVA - Open software platform for visual analytics

//Copyright (c) 2018, VTT Technical Research Centre of Finland Ltd
//All rights reserved.
//Redistribution and use in source and binary forms, with or without
//modification, are permitted provided that the following conditions are met:

//1) Redistributions of source code must retain the above copyright
//notice, this list of conditions and the following disclaimer.
//2) Redistributions in binary form must reproduce the above copyright
//notice, this list of conditions and the following disclaimer in the
//documentation and/or other materials provided with the distribution.
//3) Neither the name of the VTT Technical Research Centre of Finland nor the
//names of its contributors may be used to endorse or promote products
//derived from this software without specific prior written permission.

//THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS'' AND ANY
//EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE LIABLE FOR ANY
//DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
//(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
//LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
//ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
//(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
// @author Pekka Siltanen

function trimForScatterplot(response){
	
    if ("error" in response) {
        console.log(response);
        return(response);
    }   
    var data = new Object;    
     
    var xCoords = [];
    var yCoords = [];
    var meta = [];
    var objectids = [];
    var taskids = [];
    
    var dataMatrix = response.matrix;
    for (i in dataMatrix) {
        if (dataMatrix[i].length == 2) {
            // data frame may have missing values!!!!
            xCoords.push(parseFloat(dataMatrix[i][0]));
            yCoords.push(parseFloat(dataMatrix[i][1]));
        }
    }
     
    data["xCoords"] =  xCoords;
    data["yCoords"] =  yCoords;  
    for (var key in response) {
    	addProperty(key,data,response[key]);
    }
    
    
    return data;
}

function addProperty(key,data,value) {
	if (value != null && value!="NA" && key !="image" && key!="matrix") {
		data[key] = value;
	}
}



function minLimit(min,lowerLimit, alarmLowerLimit) {
	if (lowerLimit == null && alarmLowerLimit == null) {
		return min;
	}

	if (lowerLimit == null) {
		return Math.min(min,alarmLowerLimit)
	}

	if (alarmLowerLimit == null) {
		return Math.min(min,lowerLimit)
	}	

	return Math.min(min,lowerLimit,alarmLowerLimit)
}

function maxLimit(max,upperLimit, alarmUpperLimit) {
	if (upperLimit == null && alarmUpperLimit == null) {
		return max;
	}

	if (upperLimit == null) {
		return Math.max(max,alarmUpperLimit)
	}

	if (alarmUpperLimit == null) {
		return Math.max(max,upperLimit)
	}	

	return Math.max(max,upperLimit,alarmUpperLimit)
}


function scatterplotChart(responsedata,window,id) {
	responsedata = trimForScatterplot(responsedata);
	
	var canvasSize = getCanvasSize();	
	var margin = canvasSize.margin;
	var window = canvasSize.window;
    var w = window.width;
    var h = window.height;
	
    var mainTitle = responsedata.title;
 
    var xCoords = responsedata.xCoords;
    var yCoords = responsedata.yCoords;
    var xTitle = responsedata.xTitle;
    var yTitle = responsedata.yTitle;
 
 
    var slope = parseFloat(responsedata.slope);
    var intercept = parseFloat(responsedata.intercept);
     
    var color = "black";
 
    var innerPad = 5;
 
    var data = []
    for (i in xCoords) {
        data.push({x:parseFloat(xCoords[i]), y:parseFloat(yCoords[i])});                
    }
    
    var xExt = d3.extent(xCoords);
    var xMin = minLimit(xExt[0],responsedata.xLowerLimit, responsedata.xAlarmLowerLimit); 
    var xMax = maxLimit(xExt[1],responsedata.xUpperLimit, responsedata.xAlarmUpperLimit); 
    var yExt = d3.extent(yCoords);
    var yMin = minLimit(yExt[0],responsedata.yLowerLimit, responsedata.yAlarmLowerLimit); 
    var yMax = maxLimit(yExt[1],responsedata.yUpperLimit, responsedata.yAlarmUpperLimit); 
    var yPadding = (yMax-yMin)*0.02;
    var xPadding = (xMax-xMin)*0.02;
    
	var svg2 = addSvgRoot(id,w,h);
    
    addMainTitle(mainTitle,svg2,w,margin,0,"mainTitle");

    var scatterplot = svg2.append("g")
    .attr("id", "scatterplot")
    .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
 
//  this is a clipath for clipping the regression line
//  i.e showing only the part of the line that is within the graph
    addClipPath(scatterplot, w - (margin.left + margin.right),h -(margin.top + margin.bottom),xPadding,yPadding);
 
    var xScale, yScale;
    d3.selectAll("circle.points").remove();
    d3.selectAll("text.axes").remove();
    d3.selectAll("line.axes").remove();
     
    
    xScale = d3.scaleLinear().domain([xMin,xMax]).range([0, w - (margin.left + margin.right)]);
    yScale = d3.scaleLinear().domain([yMin,yMax]).range([h -(margin.top + margin.bottom), 0]);
        
      
    
    addXAxisTitle(scatterplot, w, h, margin, xTitle,"xAxisTitle");   
    addYAxisTitle(scatterplot, w, h, margin, yTitle); 
 
    addMainRect(scatterplot, w - (margin.left + margin.right),h -(margin.top + margin.bottom), xScale(xPadding),h  -(margin.top + margin.bottom)- yScale(yPadding));
//    xticks = xScale.ticks(5);
//    yticks = yScale.ticks(5);
 
 //   addXAxes(scatterplot,xticks,xScale,h,margin);
    addXAxisNumericScale(scatterplot,d3.axisBottom(xScale),yScale(0 - yPadding));
	addYAxis(scatterplot,d3.axisLeft(yScale), xScale(0 - xPadding));
   // addYAxes(scatterplot,yticks,yScale,w,margin);
    

//    addXGridlines(scatterplot,xticks,xScale,h,margin);  
//    addYGridlines(scatterplot,yticks,yScale,w,margin);
 
 
     
    var points = scatterplot.selectAll("empty")
    .data(d3.range(xCoords.length))
    .enter()
    .append("circle")
    .attr("class", "points")
    .attr("cx", function(d) { return xScale(xCoords[d]);})
    .attr("cy", function(d) { return yScale(yCoords[d]);})
    .attr("r", 3)
    .attr("stroke", "black")
    .attr("stroke-width", 1)
    .attr("fill", color) 
    .attr("pointer-events", "all") 
    .on("mouseover", function(d) {
        d3.select(this).attr("fill", "white");
        return tip.show(d,this);  
    }).on("mouseout", function() {
         d3.select(this).attr("fill", color);
        return tip.hide();
    });
 
    var maxWidth = xScale.invert(w - (margin.right + margin.left));
     
    // this adds a regression line y = slope*x + intercept  
     if (intercept != null && slope != null) {
    	 var x0 = xScale.invert(0);
    	 addLine(scatterplot,xScale,yScale,x0,maxWidth,intercept,intercept + slope*maxWidth,"line-sc-regression",null);
    }
    
    // add an alarm lowerlimit line
    if (responsedata.xAlarmLowerLimit != null) {
    	 addLine(scatterplot,xScale,yScale,responsedata.xAlarmLowerLimit,responsedata.xAlarmLowerLimit,yMax + yPadding ,yMin - yPadding,"line-sc-alarmLowerlimit",null);
    }
  
    if (responsedata.xLowerLimit != null) {
    	addLine(scatterplot,xScale,yScale,responsedata.xLowerLimit,responsedata.xLowerLimit,yMax + yPadding,yMin- yPadding,"line-sc-lowerLimit",null);
    } 
    
    if (responsedata.xAlarmUpperLimit != null) {
    	addLine(scatterplot,xScale,yScale,responsedata.xAlarmUpperLimit,responsedata.xAlarmUpperLimit,yMax + yPadding,yMin - yPadding,"line-sc-alarmUpperLimit",null);
    }
  
    if (responsedata.xUpperLimit != null) {
    	addLine(scatterplot,xScale,yScale,responsedata.xUpperLimit,responsedata.xUpperLimit,yMax + yPadding,yMin- yPadding,"line-sc-upperLimit",null);
    } 
   
    if (responsedata.yAlarmLowerLimit != null) {
    	addLine(scatterplot,xScale,yScale,xMin - xPadding,xMax + xPadding,responsedata.yAlarmLowerLimit,responsedata.yAlarmLowerLimit,"line-sc-alarmLowerLimit",null);
    }
  
    if (responsedata.yLowerLimit != null) {
    	addLine(scatterplot,xScale,yScale,xMin - xPadding,xMax + xPadding,responsedata.yLowerLimit,responsedata.yLowerLimit,"line-sc-lowerLimit",null);
    }
    
    if (responsedata.yAlarmUpperLimit != null) {
    	addLine(scatterplot,xScale,yScale,xMin - xPadding,xMax + xPadding,responsedata.yAlarmUpperLimit,responsedata.yAlarmUpperLimit,"line-sc-alarmUpperLimit",null);
    }
    if (responsedata.yUpperLimit != null) {
    	addLine(scatterplot,xScale,yScale,xMin - xPadding,xMax+ xPadding,responsedata.yUpperLimit,responsedata.yUpperLimit,"line-sc-upperLimit",null);
    }

//  var brush = d3.brush()
//  svg2.append('g')
//     .attr('class', 'brush')
//     .call(brush); 
//    brush.on('brush', function() {
//        var extent = brush.extent();
//        points.classed("selected", function(d) {    
//                d3.select(this).attr("fill", "black")
//          });
//    }).on("end", function() {
//         if (!d3.event.sourceEvent) return; // Only transition after input.
//         if (!d3.event.selection) return; // Ignore empty selections.
//           
//        var extent = d3.event.selection;
//         
//         
//        var minx = xScale2.invert(extent[0][0] - margin.left) ;
//        var maxx = xScale2.invert(extent[1][0] - margin.left) ;
//        var miny = yScale2.invert(extent[1][1] - margin.top)  ;
//        var maxy = yScale2.invert(extent[0][1] - margin.top) ;
//         
//        console.log(minx + " " + maxx + " " + miny + " " + maxy);
//        var oiIds = [];
//        var taskIds = [];
//        points.classed("selected", function(d) {    
//            is_brushed = minx <= data[d].x && data[d].x <= maxx && 
//                         miny <= data[d].y && data[d].y <= maxy ;
//            if (is_brushed) {
//                d3.select(this).attr("fill", "white")
//                console.log(" objectid " + data[d].objectid + " taskid " + data[d].taskid);
//                taskIds.push(data[d].taskid);   
//            }
//          });
//        makeOiList(taskIds);
//    });
     
    var tip = d3.tip()
    .attr('class', 'd3-tip')
    .offset([-10, 0])
    .html(function(d) {
        console.log(d);
        return "<span style='color:black'>" + xCoords[d] + " " + yCoords[d] + "</span>";
    })
 
    svg2.call(tip);
    var tempt = d3.select("#" + id + "> div").node();
	return d3.select("#" + id + "> div").node();
	
};

//function addXAxes(chart, xticks,xScale,h,margin) {
//	chart.selectAll("empty")
//    .data(xticks)
//    .enter()
//    .append("text")
//    .attr("class", "axes")
//    .text(function(d) { return d3.format("d")(d);})  // this shows the value in integer
//    .attr("x", function(d) { return xScale(d);})
//    .attr("y", h + margin.bottom * 0.3)
//    .attr("dominant-baseline", "middle")
//    .attr("text-anchor", "middle");
//}
    
function addXGridlines(graph,xticks,xScale,h,margin) {
	graph.selectAll("empty")
	.data(xticks)
	.enter()
	.append("line")
	.attr("class", "axes")
	.attr("x1", function(d) { return xScale(d);})
	.attr("x2", function(d) {return xScale(d);})
	.attr("y1", 0)
	.attr("y2", h)
	.attr("stroke", "LightGrey")
	.attr("stroke-width", 1);
}



//function addYAxes(chart, yticks,yScale,w,margin) {
//	chart.selectAll("empty")
//	.data(yticks)
//	.enter()
//	.append("text")
//	.attr("class", "axes")
//	.text(function(d) { return d3.format("d")(d);})
//	.attr("x", -margin.left * 0.1)
//	.attr("y", function(d) {return yScale(d);})
//	.attr("dominant-baseline", "middle")
//	.attr("text-anchor", "end");
//}

function addYGridlines(graph,yticks,yScale,w,margin) {
	graph.selectAll("empty")
	.data(yticks)
	.enter()
	.append("line")
	.attr("class", "axes")
	.attr("y1", function(d) { return yScale(d);})
	.attr("y2", function(d) { return yScale(d);})
	.attr("x1", 0)
	.attr("x2", w)
	.attr("stroke", "LightGrey")
	.attr("stroke-width", 1);
}



//function addSCLine(scatterplot,xscale,yscale,x0,x1,y0,y1,color) {
//    scatterplot
//    .append("g")
//    .attr("clip-path", "url(#clip)")
//    .append("line")
//    .attr("stroke", color)
//    .attr("stroke-width", 1)
//    .attr("x1",  xscale(x0))
//    .attr("x2", xscale(x1))
//    .attr("y1", yscale(y0))
//    .attr("y2", yscale(y1))
//    .attr("clip-path", "url(#mainrect)") 
//}



//this is a clipath for clipping the regression line
//i.e showing only the part of the line that is within the graph
function addClipPath(graph,w,h,xPadding,yPadding) {
	graph.append("svg:clipPath")
	.attr("id", "clip")
	.append("svg:rect")
	.attr("height", h)
	.attr("width", w)
	.attr("fill", d3.rgb(255,255,255))
	.attr("stroke", "black")
	.attr("stroke-width", 1)
	.attr("pointer-events", "none")
	.attr("transform", "translate(" + -xPadding +"," + -yPadding + ")");;  
}


function addMainRect(graph,w,h,xPadding,yPadding) {
	graph.append("rect")
	.attr("class","mainrect")
	.attr("height", h + 2*yPadding)
	.attr("width", w + 2*xPadding)
	.attr("fill", "none")
	.attr("stroke", "black")
	.attr("stroke-width", 1)
	.attr("pointer-events", "none")
	.attr("transform", "translate(" + -xPadding +"," + -yPadding + ")");
}

File: src/main/resources/static/R/Scatterplot.R
# OpenVA - Open software platform for visual analytics
#
# Copyright (c) 2018, VTT Technical Research Centre of Finland Ltd
# All rights reserved.
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 
# 1) Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
# 2) Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# 3) Neither the name of the VTT Technical Research Centre of Finland nor the
# names of its contributors may be used to endorse or promote products
# derived from this software without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# Author: Paula Järvinen, Pekka Siltanen

args <- commandArgs(trailingOnly = F)
scriptPath <- dirname(sub("--file=","",args[grep("--file",args)]))
print(scriptPath)

source(paste(scriptPath, "common.R", sep="/"))

connectDB <- function()
{
	Sys.setenv("TZ"="UTC")
	psql <- dbDriver("PostgreSQL")
	dbcon <- dbConnect(psql, host="<host>", port=<port>, dbname="<dbname>",user="<user>",pass="<password>")
	return(dbcon)
}

deploy_scatplot_fast=function(varids,oitype,oiids,starttime=NULL,endtime=NULL,time_unit,dbcon,imagetype)
{ 
    on.exit(dbDisconnect(dbcon))
    oiids_temp=paste("(",oiids,")",sep="")
    varidstemp=as.list(strsplit(varids, ",")[[1]])
    variableid1=(varidstemp[[1]])
    variableid2=varidstemp[[2]]
    my_meta1=getMetadata(dbcon,variableid1)
    my_meta2=getMetadata(dbcon,variableid2)

    my_ois=getOIs(dbcon,oiids_temp)
    my_ois_title=paste(my_ois[,c("report_title")],collapse = ',')

    var1_frame = getVariableValues_partition_aggregated_by_time(dbcon,variableid=variableid1, oiids=oiids,oitype=oitype,starttime=starttime,endtime=endtime,fast=FALSE, time_unit,row_limit=NULL) 
    if(nrow(var1_frame)==0) {stop(paste("OpenVA warning: No data found",my_ois_title,my_meta1$report_title) )}
    var1_frame=var1_frame[complete.cases(var1_frame),]
    zoo_frame=zoo(var1_frame$measurement_value,var1_frame$measurement_time)
    timeunit=paste(time_unit,"s",sep="")
    plot_zoo=period.apply(zoo_frame, endpoints(zoo_frame, timeunit), my_meta1$plottype)
    plot_zoo <- na.omit(plot_zoo)
    zoo_frame=fortify.zoo(plot_zoo)
    colnames(zoo_frame)=c("measurement_time","measurement_value")
    plot_frame1=zoo_frame
    

    var2_frame = getVariableValues_partition_aggregated_by_time(dbcon,variableid=variableid2, oiids=oiids,oitype=oitype,starttime=starttime,endtime=endtime,fast=FALSE, time_unit,row_limit=NULL) 
    if(nrow(var2_frame)==0) {stop(paste("OpenVA warning: No data found",my_ois_title,my_meta2$report_title) )}
    var2_frame=var2_frame[complete.cases(var2_frame),]
    zoo_frame=zoo(var2_frame$measurement_value,var2_frame$measurement_time)
    timeunit=paste(time_unit,"s",sep="")
    plot_zoo=period.apply(zoo_frame, endpoints(zoo_frame, timeunit), my_meta2$plottype)
    plot_zoo <- na.omit(plot_zoo)
    zoo_frame=fortify.zoo(plot_zoo)
    colnames(zoo_frame)=c("measurement_time","measurement_value")
    plot_frame2=zoo_frame
  
    
#create ts objects with righ timeunit
    rm(var1_frame1)
    gc()
    plot_time_unit_title= paste(my_meta2$plottype,"per",time_unit)
    rm(var2_frame2)
    gc()
    #total_sum=return_list[[3]] 
     
#merge variables   
   plot_frame=merge(plot_frame1,plot_frame2, by="measurement_time")
    
    starttime = min(min(plot_frame1$measurement_time),min(plot_frame2$measurement_time))
    endtime = max(max(plot_frame1$measurement_time),max(plot_frame2$measurement_time))
    rm(plot_frame1)
    rm(plot_frame2)
    gc()
  
  if (nrow(plot_frame)==0) {stop(paste("no joint data found: ",variableid1,variableid2,"oiids",oiids) )}
  
  colnames(plot_frame)=c("measurement_time","measurement_value1","measurement_value2")

  if ( length(unique(plot_frame$measurement_value1))==1 | length(unique(plot_frame$measurement_value2))==1) 
  { 
    stop("OpenVA warning: Not enough variation in observations for scatterplot")}


#plotting
    n=nrow(plot_frame)
#    if (nrow(plot_frame) > 29000 && imagetype == "vector") {
#        dev.off()
#    } 
    
    
    my_cor=round(cor(plot_frame$measurement_value1,plot_frame$measurement_value2,  method = "pearson", use = "complete.obs"),2)
    my_title=paste("\nCorrelation",my_meta1$report_title,"-",my_meta2$report_title,"\n",my_ois_title,"\n n=",n,"\n", starttime," - ",endtime,"\n")

    my_meta1_lowerlimit= min(plot_frame$measurement_value1)
    if (is.null(my_meta1$lowerlimit)) my_meta1$lowerlimit=NA
    if (!is.na(my_meta1$lowerlimit) )  my_meta1_lowerlimit=my_meta1$lowerlimit
        
    my_meta1_lower_alarmlimit= min(plot_frame$measurement_value1)
    if (is.null(my_meta1$lower_alarmlimit)) my_meta1$lower_alarmlimit=NA
    if (!is.na(my_meta1$lower_alarmlimit))  my_meta1_lower_alarmlimit=my_meta1$lower_alarmlimit
        
    my_meta1_upperlimit= max(plot_frame$measurement_value1)
    if (is.null(my_meta1$upperlimit) ) my_meta1$upperlimit=NA
    if (!is.na(my_meta1$upperlimit) ) my_meta1_upperlimit=my_meta1$upperlimit
        
    my_meta1_upper_alarmlimit= max(plot_frame$measurement_value1)
    if (is.null(my_meta1$upper_alarmlimit)) my_meta1$upper_alarmlimit=NA
    if (!is.na(my_meta1$upper_alarmlimit)) my_meta1_upper_alarmlimit=my_meta1$upper_alarmlimit
        
    my_meta2_lowerlimit= min(plot_frame$measurement_value2)
    if (is.null(my_meta2$lowerlimit) ) my_meta2$lowerlimit=NA
    if (!is.na(my_meta2$lowerlimit) ) my_meta2_lowerlimit=my_meta2$lowerlimit
        
    my_meta2_lower_alarmlimit= min(plot_frame$measurement_value2)
    if (is.null(my_meta2$lower_alarmlimit))  my_meta2$lower_alarmlimit=NA
    if (!is.na(my_meta2$lower_alarmlimit)) my_meta2_lower_alarmlimit=my_meta2$lower_alarmlimit
        
    my_meta2_upperlimit= max(plot_frame$measurement_value2)
    if (is.null(my_meta2$upperlimit) ) my_meta2$upperlimit =NA
    if (!is.na(my_meta2$upperlimit) )  my_meta2_upperlimit=my_meta2$upperlimit
        
    my_meta2_upper_alarmlimit= min(plot_frame$measurement_value2)
    if (is.null(my_meta2$upper_alarmlimit) ) my_meta2$upper_alarmlimit=NA
    if (!is.na(my_meta2$upper_alarmlimit) )  my_meta2_upper_alarmlimit=my_meta2$upper_alarmlimit
        
    xlim_lowend=min(min(plot_frame$measurement_value1), my_meta1_lowerlimit, my_meta1_lower_alarmlimit)
    xlim_lowend=0.9*xlim_lowend
        
    xlim_uppend=max(max(plot_frame$measurement_value1), my_meta1_upperlimit, my_meta1_upper_alarmlimit)
    xlim_uppend=1.1*xlim_uppend
     
    ylim_lowend=min(min(plot_frame$measurement_value2), my_meta2_lowerlimit, my_meta2_lower_alarmlimit)
    ylim_lowend=ylim_lowend-(ylim_lowend/10)
    ylim_uppend=max(max(plot_frame$measurement_value2), my_meta2_upperlimit, my_meta2_upper_alarmlimit)
    ylim_uppend=ylim_uppend+(ylim_uppend/10)
        
    plot(plot_frame$measurement_value1,plot_frame$measurement_value2,
            xlab=my_meta1$report_title, ylab=my_meta2$report_title, 
            xlim=c(xlim_lowend,xlim_uppend),ylim=c(ylim_lowend,ylim_uppend), 
            main=my_title, sub=paste(plot_time_unit_title,"r=",my_cor),
            cex.main=0.9)

      abline(v=my_meta1$alarm_lowerlimit,col="red",lwd=2)
      #print(meta$alarm_lowerlimit)
      abline(v=my_meta1$alarm_upperlimit,col="red",lwd=2)
      #print(meta$alarm_upperlimit)
      abline(v=my_meta1$lowerlimit,col="black",lwd=1)
      abline(v=my_meta1$upperlimit,col="black",lwd=1)
      abline(h=my_meta2$alarm_lowerlimit,col="red",lwd=2)
      print(my_meta2$alarm_lowerlimit)
      abline(h=my_meta2$alarm_upperlimit,col="red",lwd=2)
      abline(h=my_meta2$lowerlimit,col="black",lwd=1)
      abline(h=my_meta2$upperlimit,col="black",lwd=1)
    abline(lm(plot_frame$measurement_value2~plot_frame$measurement_value1), col="blue") # may not work if not much data, need to make some kind of check

    if (imagetype == "raster") {
        return_list <- list("No data returned for raster",imagetype)
        names(return_list) <-c("info","imagetype")
        return(return_list) 
    } else if (imagetype == "vector") {
        if (nrow(plot_frame) > 50000) {
                return_list <- list("Too many data values for vector visualization: showing raster instead. \n You may want to use bigger time unit!",nrow(plot_frame),"raster")
                names(return_list) <-c("warning","n","imagetype")
                return(return_list)
        }  else {
                return_list <- list("No data returned for vector",imagetype)
                names(return_list) <-c("info","imagetype")
                return(return_list) 
        }
    } else {
        if (nrow(plot_frame) > 29000) {
            return_list <- list("Too many data values for interactive visualization: showing raster instead. \n You may want to use bigger time unit!",nrow(plot_frame),"raster")
            names(return_list) <-c("warning","n","imagetype")
            return(return_list)
        }  else {
            reg <- lm(plot_frame$measurement_value2~plot_frame$measurement_value1)
            intercept <-as.character(reg[1]$coefficients[1])
            slope <- as.character(reg[1]$coefficients[2])
            print(intercept)
            print(slope)
            
            colnames( plot_frame)=c("measurement_time",my_meta1$title,my_meta2$title)

            return_matrix=as.matrix(plot_frame[,c(2,3)])
            colnames(return_matrix) <- NULL
    		rownames(return_matrix) <- NULL
    		return_matrix = apply(return_matrix, 1, as.list)

            return_list <- list(return_matrix,intercept,slope,my_title,my_ois_title,my_meta1$title,my_meta2$title,my_meta1$alarm_lowerlimit, my_meta1$alarm_upperlimit,my_meta2$alarm_lowerlimit, my_meta2$alarm_upperlimit,my_meta1$lowerlimit, my_meta1$upperlimit,my_meta2$lowerlimit, my_meta2$upperlimit,"interactive")
            names(return_list) <-c("matrix","intercept","slope","title","ois_title","xTitle","yTitle","xAlarmLowerLimit","xAlarmUpperLimit","yAlarmLowerLimit","yAlarmUpperLimit","xLowerLimit","xUpperLimit","yLowerLimit","yUpperLimit","imagetype")
            rm(plot_frame)
            gc()
            return(return_list)
        }
    }
  
}

varids = "<varids>"
oiids = "<oiids>"
oitype = "<oitype>"
imagetype = "<imagetype>"
starttime = "<starttime>"
endtime = "<endtime>"
timeunit = "<timeunit>"
outputfile = "<outputfile>"
localResultFile = "<localResultFile>"
resultUrl = "<resultUrl>"

if (timeunit =="null") {
	stop("Select time unit and aggregation operation")
}

if (starttime=="null")  {
	starttime=NULL;
}
if (endtime=="null")  {
	endtime=NULL;
}

output_data <- tryCatch(
		{			
			suppressWarnings(
					{
						if (imagetype == "vector") {
        					svglite(localResultFile);
    					} else {
							png(filename=localResultFile, width = 600, height = 600)
						}
						dbcon <- connectDB()
						data <- deploy_scatplot_fast(varids,oitype,oiids,starttime,endtime,timeunit,dbcon,imagetype)
						data["image"] = resultUrl
						data["title"] = "Scatter plot"
						data["width"] = 600
    					data["height"] = 600						
						data
					}
			)
		},
		error=function(cond) {
			message(cond)
			return_list <- list(cond$message)
			names(return_list) <-c("error")
			return(return_list)
		},
		finally={
			dev.off()
		})
write(toJSON(output_data), file=outputfile)